src/modules/twap/updaters/PartOrdersUpdater.tsx

```diff
@@ -0,0 +1,119 @@
+import { useAtomValue, useUpdateAtom } from 'jotai/utils'
+import { useEffect } from 'react'
+
+import { Order } from '@cowprotocol/contracts'
+import { OrderParameters, SupportedChainId } from '@cowprotocol/cow-sdk'
+
+import { isTruthy } from 'legacy/utils/misc'
+
+import { useWalletInfo } from 'modules/wallet'
+
+import { computeOrderUid } from 'utils/orderUtils/computeOrderUid'
+
+import { twapOrdersListAtom } from '../state/twapOrdersListAtom'
+import { TwapPartOrderItem, twapPartOrdersAtom } from '../state/twapPartOrdersAtom'
+import { TwapOrderItem } from '../types'
+
+export function PartOrdersUpdater() {
+  const { chainId, account } = useWalletInfo()
+  const twapOrdersList = useAtomValue(twapOrdersListAtom)
+  const updateTwapPartOrders = useUpdateAtom(twapPartOrdersAtom)
+
+  useEffect(() => {
+    if (!chainId || !account) return
+
+    const twapOrders = Object.values(twapOrdersList)
+
+    const ordersParts$ = twapOrders.map((twapOrder) => {
+      return generateTwapOrderParts(twapOrder, account.toLowerCase(), chainId)
+    })
+
+    Promise.all(ordersParts$).then((ordersParts) => {
+      const ordersMap = ordersParts.reduce((acc, item) => {
+        return {
+          ...acc,
+          ...item,
+        }
+      }, {})
+
+      updateTwapPartOrders(ordersMap)
+    })
+  }, [chainId, account, twapOrdersList, updateTwapPartOrders])
+
+  return null
+}
+
+async function generateTwapOrderParts(
+  twapOrder: TwapOrderItem,
+  safeAddress: string,
+  chainId: SupportedChainId
+): Promise<{ [id: string]: TwapPartOrderItem[] }> {
+  const twapOrderId = twapOrder.id
+
+  const parts = [...new Array(twapOrder.order.n)]
+    .map((_, index) => createPartOrderFromParent(twapOrder, index))
+    .filter(isTruthy)
+
+  const ids = await Promise.all(parts.map((part) => computeOrderUid(chainId, safeAddress, part as Order)))
```

Comment:
do we really need to cast the part into an order?

Reply:
Unfortunately yes. Because `computeOrderUid` is based on `@cowprotocol/contracts` but, in the rest of CowSwap we mostly use types from `@cowprotocol/cow-sdk`.

Reply:
Unfortunately yes. Because `computeOrderUid` is based on `@cowprotocol/contracts` but, in the rest of CowSwap we mostly use types from `@cowprotocol/cow-sdk`.